packages/blade/src/components/Button/BaseButton/BaseButton.tsx

```diff
@@ -0,0 +1,259 @@
+import type { ReactElement } from 'react';
+import styled from 'styled-components';
+import getIn from '../../../utils/getIn';
+import type { BaseTextProps } from '../../Typography/BaseText';
+import BaseText from '../../Typography/BaseText';
+import type { Theme } from '../../BladeProvider';
+import { useTheme } from '../../BladeProvider';
+import type { IconComponent, IconProps, IconSize } from '../../Icons';
+import makeSpace from '../../../utils/makeSpace';
+import type { TypographyPlatforms } from '../../../tokens/global/typography';
+import makeBorderSize from '../../../utils/makeBorderSize';
+import StyledBaseButton from './StyledBaseButton';
+import type { ButtonMinHeight, ButtonTypography } from './buttonTokens';
+import { typography } from './buttonTokens';
+
+export type BaseButtonIcon = undefined | null | IconComponent;
+
+type BaseButtonCommonProps = {
+  variant?: 'primary' | 'secondary' | 'tertiary' | 'link';
+  intent?: 'positive' | 'negative' | 'notice' | 'info' | 'neutral';
+  contrast?: 'low' | 'high';
+  size?: 'large' | 'medium' | 'small' | 'xsmall';
+  iconPosition?: 'left' | 'right';
+  isDisabled?: boolean;
+  isFullWidth?: boolean;
+  onClick?: () => void;
+  type?: 'button' | 'reset' | 'submit';
+};
+
+type BaseButtonWithoutIconProps = BaseButtonCommonProps & {
+  icon?: undefined;
+  children: string;
+};
+
+type BaseButtonWithIconProps = BaseButtonCommonProps & {
+  icon: IconComponent;
+  children?: string;
+};
+
+export type BaseButtonProps = BaseButtonWithIconProps | BaseButtonWithoutIconProps;
+
+const StyledBaseText = styled(BaseText)(
+  ({
+    hasIcon,
+    iconPosition,
+    iconSpacing,
+  }: BaseTextProps &
+    Pick<BaseButtonProps, 'iconPosition'> & { hasIcon: boolean; iconSpacing: string }) => ({
+    paddingLeft: hasIcon && iconPosition === 'left' ? iconSpacing : '0px',
+    paddingRight: hasIcon && iconPosition === 'right' ? iconSpacing : '0px',
+  }),
+);
+
+type BaseButtonStyleProps = {
+  iconSize: IconSize;
+  fontSize: keyof Theme['typography']['fonts']['size'];
+  lineHeight: keyof Theme['typography']['lineHeights'];
+  minHeight: ButtonMinHeight;
+  iconSpacing: string;
```

Comment:
Because the value of `iconSpacing` would be something like `4px`. This is why I had put comments earlier that some of these props are raw values and not token values. We need raw values in place where we're directly using it with a styled component. We use token values to pass it down to a blade component like `BaseText`.

https://github.com/razorpay/blade/pull/482/files#diff-7cae5a72597e5debaf6982c2f7afd5d7e4a19abbdc3e598a8e16c772a1dce75dR95

Reply:
Yeah, we should have stricter types. I've added this commit [`2851947` (#482)](https://github.com/razorpay/blade/pull/482/commits/28519470a7b186e809b118fa25a019374ef4b124) that increases type safety for most of these. Can you see if this makes sense? Had to change the type of tokens a bit too